### PR TITLE
Adds Airplay Implementation

### PR DESCRIPTION
**Pull request by redreceipt**
DESCRIPTION
-----------

kapture 2018-12-10 at 16 16 45

### What does this PR do, or why is it needed?

This gives the user the ability to stream via Apple Airplay and Google Cast.

### What automated tests did you write?

TODO:
-----

- [ ] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [#250]
- [ ] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [ ] Set a relevant reviewer

REVIEW:
-------

**Manual QA**

- [x] Manual QA on iOS and ensure it looks/behaves as expected
- [x] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [x] Read through the "Files changed" tab _very carefully_
- [x] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [x] Could the code be simpler?
- [x] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by conrad-vanl** [comment]
image

😲 

Looks like something needs to be added to the gitignore?

**Comment by redreceipt** [comment]
Maaaaaybe

**Comment by redreceipt** [comment]
A lot of work to be done, but it works!

File: packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js [hunk 1]
```diff
@@ -135,11 +157,21 @@ class FullscreenControls extends PureComponent {
...
   componentWillUnmount() {
     this.backHandler.remove();
     if (this.closeTimeout) clearTimeout(this.closeTimeout);
+
+    // remove airplay handlers
+    this.airPlayConnected.remove();
+    this.airPlayAvailable.remove();
+
+    // TODO: when should we disconnect?
+    // AirPlay.disconnect();
```
Comment (redreceipt) [inline packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js]: @conrad-vanl when do you think is the right time to disconnect Airplay? If ever?

File: packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js [hunk 2]
```diff
@@ -244,7 +276,8 @@ class FullscreenControls extends PureComponent {
                     <Title>{get(mediaPlayer, 'currentTrack.title')}</Title>
                     <Artist>{get(mediaPlayer, 'currentTrack.artist')}</Artist>
                   </Titles>
-                  <IconSm name="empty" />
+                  <AirPlayButton />
+                  {/* <IconSm name="empty" /> */}
```
Comment (redreceipt) [inline packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js]: @burkeshartsis what's the best way to approach this UI to make it right? The icon needs a little margin on the right but it looks pretty good now (see the image in PR description). Can the button be wrapped in `IconSm`?

File: packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js (hunk 1 above)
Comment (burkeshartsis) [reply packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js]: You'll probably want to move all things airplay related to a separate component like we did with `MusicControls` and then you'll probably "render" it [here](https://github.com/ApollosProject/apollos-prototype/blob/master/packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenPlayer.js#L200). Then pass what hooks you need as props to the `FullScreenControls` so you can selectively render the AirPlay icon when airplay is available. 

Based on how `MusicControls` is setup you want your component to be mounted and unmounted similarly. So `Airplay.disconnect()` would live in the airplay component I recommended above inside that components `componentWillUnmount`.

File: packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js (hunk 2 above)
Comment (burkeshartsis) [reply packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js]: Update your branch with master and `IconSm` will use `ButtonIcon` and should look good.

File: packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js (hunk 1 above)
Comment (redreceipt) [reply packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js]: @burkeshartsis help me understand why we would use `FullScreenPlayer` at all here?

Seems confusing to put the `AirplayStuff` component in `FullscreenPlayer` and the `AirplayButton` in `FullscreenControls` when we could just put both of them in `...Controls`.

Comment (burkeshartsis) [reply packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js]: I think we'll want to schedule a talk to dive into how the media player works. The short is, `FullScreenPlayer` is what renders `FullScreenControls`. Depending on our approach we may want to move the Airplay stuff higher up the stack or likely just into it's own component like we do with `MusicConrols` in which case we'll need to manage it's visibility in state higher up the chain.

File: packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js (hunk 2 above)
Comment (redreceipt) [reply packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js]: When I wrap it in `IconSm` it doesn't load, is there something I need to do to use it?

**Comment by redreceipt** [comment]
This is going to need some QA testing

File: packages/apolloschurchapp/ios/apolloschurchapp.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -1575,6 +1600,13 @@
...
 			sourceTree = BUILT_PRODUCTS_DIR;
 		};
+		DC7D3D0521A32C8500626550 /* libRNAirplay.a */ = {
+			isa = PBXReferenceProxy;
+			fileType = archive.ar;
+			path = libRNAirplay.a;
+			remoteRef = DC7D3D0421A32C8500626550 /* PBXContainerItemProxy */;
+			sourceTree = BUILT_PRODUCTS_DIR;
+		};
 /* End PBXReferenceProxy section */
```
Comment (redreceipt) [inline packages/apolloschurchapp/ios/apolloschurchapp.xcodeproj/project.pbxproj]: @burkeshartsis after you build on your machine, commit the changes so this part goes back to normal :)

File: packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js [hunk 4]
```diff
@@ -91,6 +94,9 @@ const IconLg = withTheme(({ theme }) => ({
   iconPadding: theme.sizing.baseUnit * 0.3125,
 }))(ButtonIcon);
 
+const AirPlayContainer = styled({
+  padding: 40,
```
Comment (conrad-vanl) [inline packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js]: This should use a theme value, and ideally should be an even derivative of `baseUnit`. `baseUnit` is 16px I believe, so the closest would be one of these:

```javascript
const AirPlayContainer = styled(({ theme }) => ({
  padding: theme.sizing.baseUnit * 2,
}))(View);

// ... or

const AirPlayContainer = styled(({ theme }) => ({
  padding: theme.sizing.baseUnit * 3,
}))(View);
```

File: packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js [hunk 5]
```diff
@@ -262,7 +268,10 @@ class FullscreenControls extends PureComponent {
                     <Title>{get(mediaPlayer, 'currentTrack.title')}</Title>
                     <Artist>{get(mediaPlayer, 'currentTrack.artist')}</Artist>
                   </Titles>
-                  <IconSm name="empty" disabled />
+                  {/* TODO: this is terrible */}
+                  <AirPlayContainer>
```
Comment (conrad-vanl) [inline packages/apolloschurchapp/src/ui/MediaPlayer/FullscreenControls.js]: What's keeping this from showing up on android?

**Pull request merged**